### DISCUSSION TITLE
Getting different results each run even though I am setting the random seed in UMAP model

### DISCUSSION THREAD

**Discussion by GeorgeDittmar** [MaartenGr/BERTopic#942] [Q&A]
Are there other aspects of the framework that are somewhat stochastic in nature? I set the random seed in UMAP to hopefully keep runs the same but every time I run BERTopic over my data, the number of clusters returning changes each run. Anyone else notice this or know of another seed thats needed to be set?

**Comment by MaartenGr** [comment]
If you are in the same environment each time, then there should not be any other seeds necessary. For me at least, when I set `random_state` in UMAP, the output remains exactly the same. Could you share your exact code for running into this issue?

**Reply by GeorgeDittmar** [reply]
Yeah sorry let me add the code I am using to setup the model. I am running inside a jupyter notebook on my local machine right now so  no change in environment. I did notice a weird behavior though that if I restart the notebook and rerun it will produce the same results in a series. So for example I run for the 1st time I get 100 clusters, run a second time I get 110, run a third time get 90. Then if I restart the notebook and run three times again I get the 100, 110, 90 in the same order. I would have expected it to return the 100 clusters all 3 times. Does that make sense what I am describing?

```
nr_topics = 'auto'
calc_probabilities = True
n_gram_range = (1,2)
min_cluster_size = 10
top_n_words = 10
sentence_bert_model = "../all-MiniLM-L6-v2"

# need to pull local embedding model and pass to bertopic
sentence_model = SentenceTransformer(sentence_bert_model)
umap_model = UMAP(random_state=2022)
vectorizer_model = CountVectorizer(stop_words="english")
topic_model = BERTopic(top_n_words=top_n_words,embedding_model=sentence_model, umap_model=umap_model, nr_topics=nr_topics, calculate_probabilities=calc_probabilities, n_gram_range=n_gram_range, vectorizer_model=vectorizer_model, min_topic_size=min_cluster_size)
topics, probs = topic_model.fit_transform(my_text)
```

if I rerun this cell in my notebook each time I would expect it to return the same results each run but yeah I am seeing the behavior I described above so its like somewhat deterministic but not in the way I am expecting.

**Reply by drob-xx** [reply]
I work a lot with reproducibility and in my experience your code should produce the exact same results each time. The only exception I'm aware of is with the cuML implementation of UMAP. It sounds to me like it has to be that UMAP is not producing the exact same reduction (which it should). Can you save off your embeddings, create a new UMAP instance and run UMAP.fit() against the embeddings and compare the reductions from different instances of UMAP (created with a fixed random_state)? In your case if they show that the embeddings are the same then the problem is elsewhere and if they don't then you have your culprit.

**Reply by GeorgeDittmar** [reply]
I'll try that and see as well.

**Reply by jchen04** [reply]
Hi, I'm running into similar reproducibility issues where my model output is changing even when setting the UMAP seed.  Piggybacking here because github admonished me about searching for existing discussions before starting a new thread...

My code is below.  When I look at topic 0, it's completely different between two runs (and for some reason, in one of the runs the topic has ~100 representative docs).

I did the UMAP test that drob-xx suggested above and I think the fit results it generates are consistent (by comparing `umap.graph_` across runs - let me know if there is something else I should be checking)

I tried also setting `numpy.random.seed` each time but my results were still changing between runs.

If it matters, my input docs are about ~130k tweets, preprocessed in the same way as the DTM example (https://maartengr.github.io/BERTopic/getting_started/topicsovertime/topicsovertime.html#example)

Any help would be appreciated.

```
umap_model = UMAP(random_state=42)
ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True)

sentence_model = SentenceTransformer("all-MiniLM-L6-v2")
embeddings = sentence_model.encode(df['text'].values, show_progress_bar=True)

topic_model = BERTopic(calculate_probabilities=False,
                        nr_topics='auto',
                        ctfidf_model=ctfidf_model,
                        seed_topic_list=seed_topic_list,
                        umap_model=umap_model,
                        min_topic_size=100,
                       )
topics, probs = topic_model.fit_transform(df['text'].values,
                                           embeddings
                                           )
```

**Reply by GeorgeDittmar** [reply]
Its weird my issue has gone away randomly now after a few notebook restarts so I dunno whats going on now. Still investigating myself.

**Reply by MaartenGr** [reply]
Could you share also the import that you have as well as the versions of BERTopic, including its dependencies?

**Reply by drob-xx** [reply]
The way it is supposed to work is that there is no variation as long as you use the same `random_state` and are running on the same architecture (OS, chips etc. etc. may not return the same results.) There is a bug in the cuML UMAP implementation and random_state has to be coupled with `init='random'` (which has other implications) for the results to be the same. There should be no doubt - when we say "the same" it means exactly the same. I just compare the embeddings so you should be able to just compare embeddings:

> `np.all(bertopic_model_1.umap_model.embedding_ == bertopic_model_2.umap_model.embedding_)`

This will be `True` if the UMAP fit was the same across multiple runs.

**Reply by jchen04** [reply]
@MaartenGr  These are my imports:
```
from bertopic import BERTopic
from sentence_transformers import SentenceTransformer
from hdbscan import HDBSCAN
from bertopic.vectorizers import ClassTfidfTransformer
from umap import UMAP
```
And these are the versions:
BERTopic v0.13.0
hdbscan v 0.8.29
numpy v1.21.4
pandas v1.3.4
plotly v5.11.0
pyyaml v5.4.1
scikit-learn v1.2.0
sentence-transformers v2.2.2
tqdm v4.64.1
umap-learn v0.5.3

I'm running in a jupyter notebook on a windows 10 machine and not using cuML UMAP.

@drob-xx `np.all(topic_model1.umap_model.embedding_ == topic_model2.umap_model.embedding_)` is returning `False` for me.

Just to be extra safe, I did another test where I re-initialized all the objects in each run, as well as copied my input data and embeddings into new objects (I didn't see these being changed by the model fit, but just in case...)
```
df = df_base.copy()
umap_model = UMAP(random_state=42)
hdbscan_model = HDBSCAN(min_cluster_size=50)
embeddings = embeddings_base.copy()
ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True)
np.random.seed(100)
topic_model = BERTopic(calculate_probabilities=False,
                        nr_topics='auto',
                        ctfidf_model=ctfidf_model,
                        seed_topic_list=seed_topic_list,
                        umap_model=umap_model,
                        min_topic_size=100,
                       )
topics, probs = topic_model.fit_transform(df['text'].values,
                                           embeddings
                                           )
```

I sincerely hope this is user error somehow on my part!

**Reply by drob-xx** [reply]
Not sure what is going on. I'm sure @MaartenGr will see more than I do. However, I think you are making this more complicated than it needs to be - also your approach is going to use different default parameters than BERTopic uses out of the box. I would make a more straight forward test case e.g.

- Create two different BERTopic models 
- Run them with the umap_model random_state set the same
- compare them - they should be the same - something like:

```
my_docs = df_base['My_Text_Column']

bt1 = BERTopic()
bt1.umap_model.random_state = 42
bt1.hdscan_model.min_cluster_size = 50
bt1.fit_transform(my_docs)

bt2 = BERTopic()
bt2.umap_model.random_state = 42
bt2.hdscan_model.min_cluster_size = 50
bt2.fit_transform(my_docs)

assert(bt1.umap_model.embedding_ == bt2.umap_model.embedding_)
```

All the output from both models, the labels, the topic model vocabularies etc. should be *exactly* the same.
